accord-core/src/main/java/accord/coordinate/FetchData.java

```diff
@@ -415,14 +442,83 @@ public Void apply(SafeCommandStore safeStore)
                     break;
             }
 
+            return updateDurability(safeStore, safeCommand);
+        }
+
+        // if can only propagate Truncated, we might be stale; try to upgrade the
+        private Known applyOrUpgradeTruncated(SafeCommandStore safeStore, SafeCommand safeCommand, Command command)
+        {
+            // if our peers have truncated this command, then either:
+            // 1) we have already applied it locally; 2) the command doesn't apply locally; 3) we are stale; or 4) the command is invalidated
+            if (command.saveStatus().isUninitialised())
+                return null; // TODO (expected): maybe stale?
+
+            if (command.hasBeen(PreApplied))
+            {
+                updateDurability(safeStore, safeCommand);
+                return null;
+            }
+
+            if (Infer.safeToCleanup(safeStore, command, route, full.executeAt))
+            {
+                Commands.setErased(safeStore, safeCommand);
+                return null;
+            }
+
+            // TODO (required): erasing implies that the whole command has been fully applied globally, when in fact it
+            //  might only be a single shard that has Truncated its state. The simplest thing for us here to ensure consistency
+            //  is to permit partial application where possible, instead of erasing the whole state; we could also consider
+            //  more clearly reifying the difference between Erased and Truncated(WithoutAnyKnowledge).
+            //  In practice this may not be important, because a coordinator cannot have relied on this replica's vote for durability
+            //  and so our Erased vote should not adversely impact any inferences from a quorum of responses. But for consistency
+            //  and ease of logic this is best to retain.
+            // we're now at least partially stale, but let's first see if we can progress this shard, or we can do so in part
+            Timestamp executeAt = command.executeAtIfKnown(full.executeAtIfKnown());
+            if (executeAt == null)
+            {
+                // we don't even know the execution time, so we cannot possibly proceed besides erasing the command state and marking ourselves stale
+                safeStore.commandStore().markShardStale(safeStore, txnId, route.participants().toRanges(), false);
```

Comment:
The intention is that nodes become stale only when they are marked as long-term unhealthy - it's likely this will be an operator action when a node is failing for some reason. Right now we don't actually have a mechanism for intentionally permitting a node to go stale, though it is a very simple mechanism to add and likely to form part of the operator integrations we need to deliver.